community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java

```diff
@@ -439,6 +451,10 @@ public PrimitiveIntIterator relationshipGetPropertyKeys( long relationshipId ) t
         {
             return relationship.get().getPropertyKeys();
         }
+        finally
+        {
+            statement.assertOpen();
+        }
```

Comment:
Yes. I think most of our iterators are lazy, so this is worth investigating.
The value is usually realised in the `hasNext` call of our implementations, which makes sense considering a `hasNext` that returns `true` is supposed to guarantee that `next` will return something. Thus you should be able to create wrappers that `assertOpen` as part of the `hasNext` calls.

Cheers,
Chris

> On 01 Jul 2016, at 09:48, Anton Persson <notifications@github.com> wrote:
> 
> In community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java <https://github.com/neo4j/neo4j/pull/7477#discussion_r69261723>:
> 
> > @@ -439,6 +451,10 @@ public PrimitiveIntIterator relationshipGetPropertyKeys( long relationshipId ) t
> >          {
> >              return relationship.get().getPropertyKeys();
> >          }
> > +        finally
> > +        {
> > +            statement.assertOpen();
> > +        }
> If the iterator is lazy, then yes you are correct in that it wont protect against anything.
> If it is not we have the same scenario as above. Right?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub <https://github.com/neo4j/neo4j/pull/7477/files/e50e5ab9232bbc1050ccbfb3a4fa678df8f5aa8a#r69261723>, or mute the thread <https://github.com/notifications/unsubscribe/AAAfOWz6EHIpCc4yHReGQLAjg6KZ7awjks5qRMZWgaJpZM4JBmm9>.
>